contracts/src/StabilityPool.sol

```diff
@@ -457,33 +455,39 @@ contract StabilityPool is LiquityBase, IStabilityPool, IStabilityPoolEvents {
         if (_debtToOffset == _totalBoldDeposits) {
             boldLossPerUnitStaked = DECIMAL_PRECISION; // When the Pool depletes to 0, so does each deposit
             lastBoldLossError_Offset = 0;
+            lastBoldLossError_TotalDeposits = _totalBoldDeposits;
         } else {
-            uint256 boldLossNumerator = _debtToOffset * DECIMAL_PRECISION - lastBoldLossError_Offset;
```

Comment:
I'm wondering was this right in the first place (or maybe I need to refresh).  Prior to this PR we were subtracting the error from the bold loss (L461).  But shouldn't we add it, as we do for the coll?
 
Let's consider what we were trying to fix, if there was no error correction at all:

- `boldLossPerUnitStaked` (b) is too small, due to floor division
- `newProductFactor` (f) is too big since `f = 1e18 - b`
-  Therefore `P` is not reduced enough at the liquidation
- Therefore the deposit calc,`P/P_0 * initialDeposit` is too large
- So collateral gains and BOLD gains will also be too high, as they use `P` values in numerator at each reward
- This could lead to underflows withdrawing deposits, since 

So with no error correction it's like that.

But by subtracting the last BOLD error, don't we make it worse: `b` would already be too small from the floor division, and we then make it smaller.  

In contrast, with the coll (line 452), we add the last coll error, which makes sense: coll rewards are too small from floor div, so we're storing the error and adding it.

I wonder if simply changing the sign here in L461 would fix underflow (or if its not so simple)?

Reply:
I don’t think that’s what we were trying to fix with the error. That we do it with the`+ 1` in `boldLossPerUnitStaked` assignment below. We are doing a (pseudo-) ceil division.
Then with the `lastBoldLossError_Offset` what we are trying is that the difference between the real theoretical value and the value resulting from the ceil division doesn’t grow too big over time.
So:
- With the `+ 1` we try to avoid underflows
- With `lastBoldLossError_Offset` we try that the accumulated error from the previous measure doesn’t accumulate.

The problem is that the second step can revert the effect of the first one to some extent, and make the underflows appear again (if I’m not mistaken because the treatment of that error was wrong). That’s what we are trying to fix in this PR.

Reply:
> In contrast, with the coll (line 452), we add the last coll error, which makes sense: coll rewards are too small from floor div, so we're storing the error and adding it.

Yes, coll rewards have the opposite direction because it’s a floor division (as opposed to the ceil division of bold deposits). I.e., we don’t do the `+1` trick.

Reply:
>Yes, coll rewards have the opposite direction because it’s a floor division (as opposed to the ceil division of bold deposits). I.e., we don’t do the +1 trick.

Ahhh right. Thanks, yes that makes it clear!